pkg/gardenlet/controller/networkpolicy/peers.go

```diff
@@ -16,24 +16,77 @@ package networkpolicy
 
 import (
 	"net"
+
+	networkingv1 "k8s.io/api/networking/v1"
 )
 
-// private8BitBlock returns a private network (RFC1918) 10.0.0.0/8 IPv4 block
+// private8BitBlock returns a private network (RFC1918) 10.0.0.0/8 IPv4 block.
 func private8BitBlock() *net.IPNet {
 	return &net.IPNet{IP: net.IP{10, 0, 0, 0}, Mask: net.CIDRMask(8, 32)}
 }
 
-// private12BitBlock returns a private network (RFC1918) 172.16.0.0/12 IPv4 block
+// private12BitBlock returns a private network (RFC1918) 172.16.0.0/12 IPv4 block.
 func private12BitBlock() *net.IPNet {
 	return &net.IPNet{IP: net.IP{172, 16, 0, 0}, Mask: net.CIDRMask(12, 32)}
 }
 
-// private16BitBlock returns a private network (RFC1918) 192.168.0.0/16 IPv4 block
+// private16BitBlock returns a private network (RFC1918) 192.168.0.0/16 IPv4 block.
 func private16BitBlock() *net.IPNet {
 	return &net.IPNet{IP: net.IP{192, 168, 0, 0}, Mask: net.CIDRMask(16, 32)}
 }
 
-// carrierGradeNATBlock returns a Carrier-grade NAT (RFC6598) 100.64.0.0/10 IPv4 block
+// carrierGradeNATBlock returns a Carrier-grade NAT (RFC6598) 100.64.0.0/10 IPv4 block.
 func carrierGradeNATBlock() *net.IPNet {
 	return &net.IPNet{IP: net.IP{100, 64, 0, 0}, Mask: net.CIDRMask(10, 32)}
 }
+
+// allPrivateNetworkBlocks returns a list of all Private network (RFC1918) and Carrier-grade NAT (RFC6598) IPv4 blocks.
+func allPrivateNetworkBlocks() []net.IPNet {
+	return []net.IPNet{
+		*private8BitBlock(),
+		*private12BitBlock(),
+		*private16BitBlock(),
+		*carrierGradeNATBlock(),
+	}
+}
+
+// toNetworkPolicyPeersWithExceptions returns a list of networkingv1.NetworkPolicyPeers whose ipBlock.cidr points to
+// `networks` and whose ipBlock.except points to `except`.
+func toNetworkPolicyPeersWithExceptions(networks []net.IPNet, except ...string) ([]networkingv1.NetworkPolicyPeer, error) {
+	var result []networkingv1.NetworkPolicyPeer
+
+	for _, n := range networks {
+		excluded, err := excludeBlock(&n, except...)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, networkingv1.NetworkPolicyPeer{
+			IPBlock: &networkingv1.IPBlock{
+				CIDR:   n.String(),
+				Except: excluded,
+			},
+		})
+	}
+
+	return result, nil
+}
+
+func excludeBlock(parentBlock *net.IPNet, cidrs ...string) ([]string, error) {
+	var matchedCIDRs []string
+
+	for _, cidr := range cidrs {
+		ip, ipNet, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return nil, err
+		}
+		parentBlockMaskLength, _ := parentBlock.Mask.Size()
+		ipNetMaskLength, _ := ipNet.Mask.Size()
+
+		if parentBlock.Contains(ip) && parentBlockMaskLength < ipNetMaskLength {
+			matchedCIDRs = append(matchedCIDRs, cidr)
+		}
```

Comment:
This does not work well if the `parentBlock` and `ipNet` are essentially the same.
For example, if we have a seed with a pod cidr `192.168.0.0/16`, the `allow-to-private-networks` policy should semantically exclude this range, but it will not because the ranges are equal. The same is true if any other of the private ranges is utilised to its full extend, which is less likely for the bigger ranges.
That being said, the previous implementation had the exact same issue. Therefore, you may consider this a nit.

Reply:
Yes, I just copied this over, right. If you have a suggestion for a fix, I'm happy to include it, though. :)

Reply:
Sounds reasonable 👍🏻